presto-main/src/main/java/com/facebook/presto/sql/relational/SqlToRowExpressionTranslator.java

```diff
@@ -787,6 +787,11 @@ protected RowExpression visitSubscriptExpression(SubscriptExpression node, Void
             RowExpression base = process(node.getBase(), context);
             RowExpression index = process(node.getIndex(), context);
 
+            // this block will handle row subscript, converts the ROW_CONSTRUCTOR with subscript to a DEREFERENCE expression
+            if (base instanceof SpecialFormExpression && base.getType() instanceof RowType) {
+                ConstantExpression pos = (ConstantExpression) index;
```

Comment:
I'm not sure if `index` is always constant. Check my other comment in the desugaring rule. Could you also add a test to cover that?

Reply:
for a subscript with Rowtype it should always be a constant, otherwise it has to be an array or map

Reply:
I think the following query is valid. That's why I'm concerned.
```sql
WITH index AS (
    SELECT
        *
    FROM (
        VALUES
            (1),
            (2),
            (3),
            (4)
    ) t(col)
)
SELECT
    ARRAY[1, 2][col % 2 + 1]
FROM index
```

BTW, the plan is
```
Fragment 0 [SINGLE]
    Output layout: [expr_7]
    Output partitioning: SINGLE []
    Stage Execution Strategy: UNGROUPED_EXECUTION
    - Output[_col0] => [expr_7:integer]
            Estimates: {rows: 4 (20B), cpu: 60.00, memory: 0.00, network: 0.00}
            _col0 := expr_7
        - Project[] => [expr_7:integer]
                Estimates: {rows: 4 (20B), cpu: 60.00, memory: 0.00, network: 0.00}
                expr_7 := "$literal$array(integer)"("from_base64"('CQAAAElOVF9BUlJBWQIAAAAAAQAAAAIAAAA='))[CAST((("field" % 2) + 1) AS bigint)]
            - LocalExchange[ROUND_ROBIN] () => [field:integer]
                    Estimates: {rows: 4 (20B), cpu: 40.00, memory: 0.00, network: 0.00}
                - Limit[10000] => [field:integer]
                        Estimates: {rows: 4 (20B), cpu: 20.00, memory: 0.00, network: 0.00}
                    - Values => [field:integer]
                            Estimates: {rows: 4 (20B), cpu: 0.00, memory: 0.00, network: 0.00}
                            (INTEGER 1)
                            (INTEGER 2)
                            (INTEGER 3)
                            (INTEGER 4)

Reply:
Hi, @bhhari, I don't think it is true:
```
WITH index AS (
    SELECT
        *
    FROM (
        VALUES
            (1),
            (2),
            (3),
            (4)
    ) t(col)
)
SELECT
    Row(1, 'a', true)[col % 2 + 1]
FROM index
```
will fail this check. I think the above still seem legit? If not, we should throw a SYNTAX_ERROR instead. 

Can you make sure the `Expression` is always desugared before `RowExpression` instead?
Add a function `Expression desugar(Expression expression, SymbolAllocator symbolAllocator)` for example that will carry out all `Desugaring` and make sure it happens before translation.

Reply:
I don't think row subscript is in the standard. If that's the case it's up to us to decide whether to allow it or not. In general I think it's reasonable to only allow constant subscript. Unlike array, different fields on `ROW` can be of different type. It makes no sense to allow dynamically addressing a field from a row. That sounds like asking for trouble to me.

Reply:
Also consider row subscript is only sugaring on the underlying data structure, do we allow dynamically addressing a field in row by name? This two behaviors should be consistent.

Reply:
But `Row(1, 'a', true)[1+1]` will also fail? If there is no standard, we can just allow constant for now. It is quite rare people will need to write a query like that. The original PR doesn't seem to handle it also.

Reply:
Yes, this PR handles back porting + support for Row expressions

Reply:
@hellium01 

For now I think just supporting the most straightforward case is OK :). Just ask user to manually rewrite "1+1" to "2" 😃